packages/frontend/src/contexts/unirep.js

```diff
@@ -0,0 +1,178 @@
+    this.state = state
+    if (typeof window !== 'undefined') {
+      //   this.load()
+    }
+  }
+
+  //   async load() {}
+
+  async loadAllSignUps() {
+    const url = new URL(`api/unirep/signups`, SERVER)
+    const data = await fetch(url.toString()).then((r) => r.json())
+    this.allSignUps = data
+    this.ingestSignUps(data)
+  }
+
+  async ingestSignUps(signups) {
+    this.signUpsByUserId = new Map()
+    this.signUpsByAttesterId = new Map()
+    for (let i = 0; i < signups.length; i++) {
+      let userId = signups[i].commitment
+      let attesterId = signups[i].attesterId
+      if (this.signUpsByUserId.has(userId)) {
+        this.signUpsByUserId.get(userId).push(signups[i])
+      } else {
+        this.signUpsByUserId.set(userId, [signups[i]])
+      }
+
+      if (this.signUpsByAttesterId.has(attesterId)) {
+        this.signUpsByAttesterId.get(attesterId).push(signups[i])
+      } else {
+        this.signUpsByAttesterId.set(attesterId, [signups[i]])
+      }
+    }
+    console.log('userSignups:', this.signUpsByUserId)
+    console.log('attesterSignUps:', this.signUpsByAttesterId)
+    console.log(
+      'one users signups with map.get:',
+      this.signUpsByUserId.get(
+        '21148151481457093107206483541042547669092147310094944251743153632587065177648'
+      )
+    )
+    console.log('instanceof?', this.signUpsByUserId instanceof Map)
+  }
+
+  async loadAllAttestations() {
+    const url = new URL(`api/unirep/attestations`, SERVER)
+    const data = await fetch(url.toString()).then((r) => r.json())
+    this.allAttestations = data
+    this.allAttestations.forEach((attestation) => {
+      this.totalPosRep += attestation.posRep
+      this.totalNegRep += attestation.negRep
+    })
+    // this.ingestAttestations(data)
+  }
+
+  // async ingestAttestations(attestations) {
+  //   this.attestationsByAttesterId = new Map()
+  //   for (let i = 0; i < attestations.length; i++) {
+  //     let attesterId = attestations[i].attesterId
+  //     if (this.attestationsByAttesterId.has(attesterId)) {
+  //       this.attestationsByAttesterId.get(attesterId).push(attestations[i])
+  //     } else {
+  //       this.attestationsByAttesterId.set(attesterId, [attestations[i]])
+  //     }
+  //     this.totalPosRep += attestations[i].posRep
+  //     this.totalNegRep += attestations[i].negRep
+  //   }
+  //   console.log('attestationsByAttester:', this.attestationsByAttesterId)
+  //   console.log('total rep:', this.totalNegRep, this.totalPosRep)
+  // }
+
+  async loadAllEpochs() {
+    const url = new URL(`api/unirep/epochs`, SERVER)
+    this.allEpochs = await fetch(url.toString()).then((r) => r.json())
+    this.allEpochs.forEach((epoch) => {
+      epoch.number === 0 && this.attesterIds.push(epoch.attesterId)
+    })
+    this.getCurrentAttesterStats()
+    console.log('loadAllEpochs was called')
+  }
+  // create array of attester stats
+  getCurrentAttesterStats() {
+    this.attesterIds.forEach((attester) => {
```

Comment:
I'm not a fan of this nested `forEach` logic. I think this function could be refactored to be more clear

Reply:
Instead try a reducer?

Reply:
Yes. But I'd like to leave it for now bc I'm going to approach it differently with data from the new event.